From 2c1b0e913524b1cdf1d7d99b3259e29c1a3c8d2b Mon Sep 17 00:00:00 2001
From: Nadav Haklai <nadavh@marvell.com>
Date: Tue, 9 Jun 2015 15:29:19 +0300
Subject: [PATCH 0249/2241] ARM: mm: fix support for HW coherent systems in
 PL310 cache

When a PL310 cache is used in a system that provides hardware coherency,
the entire outer cache operations are useless, and can be skipped.
Moreover, on some systems, it is harmful as it causes deadlocks between
the Marvell coherency mechanism, the Marvell PCIe controller and the
Cortex-A9.

This commit extends a previous commit:
'commit 98ea2dba6593 ("ARM: 8076/1: mm: add support for HW coherent systems
in PL310 cache")' which added the io-coherent support for the PL310 cache
by also disabling the outer cache flush range operation.

In the current kernel implementation, the outer cache flush range
operation is triggered by the dma_alloc function.
This operation can be take place during runtime and in some circumstances
may lead to the PCIe/PL310 deadlock on Armada 375/38x SoCs.

Signed-off-by: Nadav Haklai <nadavh@marvell.com>
Change-Id: I253ac4dcb323627edf59cbabad42c01650b1ce7f
Reviewed-on: http://vgitil04.il.marvell.com:8080/19932
Reviewed-by: Ofer Heifetz <oferh@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/28160
---
 arch/arm/mm/cache-l2x0.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mm/cache-l2x0.c b/arch/arm/mm/cache-l2x0.c
index 493692d..964a09a 100644
--- a/arch/arm/mm/cache-l2x0.c
+++ b/arch/arm/mm/cache-l2x0.c
@@ -1271,9 +1271,9 @@ static const struct l2c_init_data of_l2c310_data __initconst = {
 };
 
 /*
- * This is a variant of the of_l2c310_data with .sync set to
- * NULL. Outer sync operations are not needed when the system is I/O
- * coherent, and potentially harmful in certain situations (PCIe/PL310
+ * This is a variant of the of_l2c310_data with .sync and .flush_range set to
+ * NULL. Outer sync and flush range operations are not needed when the system
+ * is I/O coherent, and potentially harmful in certain situations (PCIe/PL310
  * deadlock on Armada 375/38x due to hardware I/O coherency). The
  * other operations are kept because they are infrequent (therefore do
  * not cause the deadlock in practice) and needed for secondary CPU
@@ -1292,7 +1292,6 @@ static const struct l2c_init_data of_l2c310_coherent_data __initconst = {
 	.outer_cache = {
 		.inv_range   = l2c210_inv_range,
 		.clean_range = l2c210_clean_range,
-		.flush_range = l2c210_flush_range,
 		.flush_all   = l2c210_flush_all,
 		.disable     = l2c310_disable,
 		.resume      = l2c310_resume,
-- 
2.7.4

